File: src/main/java/ru/donz/wf/behaviors/FlowBehavior.java
package ru.donz.wf.behaviors;

import akka.persistence.typed.PersistenceId;
import akka.persistence.typed.javadsl.CommandHandler;
import akka.persistence.typed.javadsl.Effect;
import akka.persistence.typed.javadsl.EventHandler;
import akka.persistence.typed.javadsl.EventSourcedBehavior;

public class FlowBehavior extends EventSourcedBehavior<FlowProtocol, FlowEvent, FlowState> {

    public FlowBehavior(PersistenceId persistenceId) {
        super(persistenceId);
    }

    @Override
    public FlowState emptyState() {
        return new FlowState("", "", "");
    }

    @Override
    public CommandHandler<FlowProtocol, FlowEvent, FlowState> commandHandler() {
        return newCommandHandlerBuilder()
                .forAnyState()
                .onCommand(FlowProtocol.IncomingExternalMessage.class, this::onIncomingExternalMessage)
                .build();
    }

    @Override
    public EventHandler<FlowState, FlowEvent> eventHandler() {
        return newEventHandlerBuilder()
                .forAnyState()
                .onEvent(FlowEvent.IncomingExternalMessage.class, FlowState::newIncomingExternalMessage)
                .build();
    }

    private Effect<FlowEvent, FlowState> onIncomingExternalMessage(FlowState currentState, FlowProtocol.IncomingExternalMessage message) {
        return Effect().persist(new FlowEvent.IncomingExternalMessage(message.getMessage()))
                .thenReply(message.getReplyTo(), s -> "OK");
    }
}

File: src/main/kotlin/ru/onix/wf/behaviors/FlowBehavior.kt
package ru.onix.wf.behaviors

import akka.persistence.typed.PersistenceId
import akka.persistence.typed.javadsl.CommandHandler
import akka.persistence.typed.javadsl.EventHandler
import akka.persistence.typed.javadsl.EventSourcedBehavior

class FlowBehavior(val persistenceId: PersistenceId) :
    EventSourcedBehavior<FlowCommand, FlowEvent, FlowState>(persistenceId) {

    override fun emptyState(): FlowState {
        return FlowState("", "", "")
    }

    override fun commandHandler(): CommandHandler<FlowCommand, FlowEvent, FlowState> {
        return newCommandHandlerBuilder()
            .forAnyState()
            .onCommand(FlowCommand.IncomingExternalMessage::class.java, this::onIncomingExternalMessage)
            .build()
    }

    override fun eventHandler(): EventHandler<FlowState, FlowEvent> {
        return newEventHandlerBuilder()
            .forAnyState()
            .onEvent(FlowEvent.IncomingExternalMessage::class.java, FlowState::newIncomingExternalMessage)
            .build()
    }

    private fun onIncomingExternalMessage(
        state: FlowState,
        message: FlowCommand.IncomingExternalMessage,
    ) = Effect()
        .persist(FlowEvent.IncomingExternalMessage(message.message))
        .thenReply(message.replyTo, { "OK" })
}